### ISSUE TITLE
Publish wasm

### ISSUE THREAD

**Issue by Boshen** [oxc-project/oxc#301]

**Comment by sxzz** [comment]
Hi, I would like to know if there is a reason to close this issue. Are there any plans to release wasm version?

**Comment by Boshen** [comment]
> Hi, I would like to know if there is a reason to close this issue. Are there any plans to release wasm version?

I haven't found the time to properly maintain the wasm version. What's your usecase?

**Comment by sxzz** [comment]
I'd like to add Oxc parser to my [AST Explorer](https://ast.sxzz.moe/)

**Comment by 0xdevalias** [comment]
I noticed that the playground's `package.json` loads it as such:

- https://github.com/oxc-project/oxc/blob/main/website/package.json#L32C5-L32C19
  - `"@oxc/wasm-web": "link:../npm/wasm-web",`

Yet there doesn't seem to be a `npm/wasm-web` in the main repo?

- https://github.com/oxc-project/oxc/tree/main/npm

**Comment by 0xdevalias** [comment]
> What's your usecase?

We were also exploring whether it would be a suitable parser in `wakaru`:

- https://github.com/pionxzh/wakaru/issues/35#issuecomment-1835413696
- https://github.com/pionxzh/wakaru/issues/35#issuecomment-1835430137

**Comment by Boshen** [comment]
> > What's your usecase?
> 
> @Boshen We were also exploring whether it would be a suitable parser in `wakaru` (cc // @pionxzh):
> 
> * [Split code and save progress? pionxzh/wakaru#35 (comment)](https://github.com/pionxzh/wakaru/issues/35#issuecomment-1835413696)
> * [Split code and save progress? pionxzh/wakaru#35 (comment)](https://github.com/pionxzh/wakaru/issues/35#issuecomment-1835430137)
> 
> > I haven't found the time to properly maintain the wasm version
> 
> @Boshen Off the top of your head, what are the current issues with it that require further maintenance/etc? Do you consider it not up to date/usable/etc?

Things could work if you intend to write everything in Rust, otherwise it's not going to work on the JS side. We lack all the things you need on the JS side after looking at your requirements.

**Comment by Boshen** [comment]
> We will need the wasm binding for js parser and the AST query and run them on JS side.
> 
> So the main blocker is that there is no enough resources and maintainers for wasm version?

> AST query and run them on JS side
We don't have a AST visitor on the JS side.

> So the main blocker is that there is no enough resources and maintainers for wasm version?

Kind of, it's gonna be a lot harder to maintain because there are currently no TypeScript types for the AST on the JS side.

**Comment by magic-akari** [comment]
> We don't have a AST visitor on the JS side.

Is using a WASM plugin a good idea?

**Comment by Boshen** [comment]
To be honest I'm still not sure whether the scope of this project should expand to supporting wasm and napi ...

**Comment by taosx** [comment]
Appreciate trying to keep the scope small but I was thinking that the the oxc ecosystem could benefit from supporting multiple use cases with its components. That way it would gain new contributors and would ease some of the maintaining burden. The community would also benefit by not needing to duplicate effort redoing lexers, parsers, traversers...etc.

My usecase involves rewriting some JS/TS from both js and external systems and oxc came to mind.

**Comment by Boshen** [comment]
Done! https://www.npmjs.com/package/@oxc-parser/wasm

I used `wasm-pack build --target web` for the build, probably need something like https://github.com/nshen/vite-plugin-wasm-pack to get it working with vite, otherwise vite will load the wasm file as a HTML file causing a` CompileError: WebAssembly.instantiate(): expected magic word` error.

**Comment by sxzz** [comment]
Thanks for your excellent work! However, the usage of docs may be wrong.

```ts
parseSync(code, { filename: "test.ts" });
> Error: expected instance of ParserOptions
```
The parser options cannot be a plain object.


---

If we pass a class, there's an error once get/set value.
```ts
const parserOptions = new ParserOptions()
parserOptions.filename = options?.filename || 'foo.ts'
> Error: null pointer passed to rust
```

**Comment by Boshen** [comment]
@sxzz I released @oxc-parser/wasm@0.0.4 and updated the README https://www.npmjs.com/package/@oxc-parser/wasm

wasm FFI is really confusing, I'm still unsure whether I fixed it or not, but "it works on my machine" at least.

**Comment by sxzz** [comment]
Great, it works now. 

P.S. Oxc parser is already supported by my [AST Explorer](https://ast.sxzz.moe) version.
